engines/glk/alan2/parse.cpp

```diff
@@ -0,0 +1,976 @@
+					if (mlst[i].code != 0 && !isObj(mlst[i].code)) // Skip any empty slots
+						mlst[i].code = 0;
+				}
+			} else if (!isObj(params[p].code))
+				_vm->printError(M_CANT0);
+		}
+	}
+
+	// Set verb code
+	_vm->cur.vrb = ((Aword *) cla)[1];	// Take first word after end of table!
+
+	// Finally, if ALL was used, try to find out what was applicable
+	if (allLength > 0) {
+		for (p = 0; params[p].code != 0; p++); // Find multiple marker
+
+		for (i = 0; i < allLength; i++) {
+			if (mlst[i].code != 0) {	// Already empty?
+				params[p] = mlst[i];
+
+				if (!possible())
+					mlst[i].code = 0;	// Remove this from list
+			}
+		}
+
+		params[p].code = 0;		// Restore multiple marker
+		listCompact(mlst);
+
+		if (listLength(mlst) == 0) {
+			params[0].code = EOF;
+			_vm->printError(M_WHAT_ALL);
+		}
+	} else if (anyPlural) {
+		listCompact(mlst);
+
+		// If there where multiple parameters but non left, exit without a
+		// word, assuming we have already said enough
+		if (listLength(mlst) == 0)
+			_vm->printError(MSGMAX);
+	}
+
+	plural = anyPlural;		// Remember that we found plural objects
+}
+
+// Find the verb class (not used currently) and 'tryMatch()'
+void Parser::match(ParamElem *mlst) {
+	tryMatch(mlst);			// try to understand what the user said
+
+	if (wrds[wrdidx] != EOF && !isConj(wrds[wrdidx]))
+	    _vm->printError(M_WHAT);
+	if (wrds[wrdidx] != EOF)	// More on this line?
+		wrdidx++;			// If so skip the AND
+}
+
+// Execute all activities commanded.Handles possible multiple actions
+// such as THEM or lists of objects.
+void Parser::action(ParamElem plst[]) {
+	int i, mpos;
+	char marker[10];
+
+	if (plural) {
+		// The code == 0 means this is a multiple position. We must loop
+		// over this position (and replace it by each present in the plst)
+		for (mpos = 0; params[mpos].code != 0; mpos++); // Find multiple position
+		
+		sprintf(marker, "($%d)", mpos + 1); // Prepare a printout with $1/2/3
+
+		for (i = 0; plst[i].code != EOF; i++) {
+			params[mpos] = plst[i];
+			_vm->output(marker);
+			//do_it();	// TODO
+
+			if (plst[i + 1].code != EOF)
+				_vm->paragraph();
+		}
+
+		params[mpos].code = 0;
+	} //else // TODO
+		//do_it();
+}
+
+void Parser::parse() {
+	if (mlst == NULL) {		// Allocate large enough paramlists
+		mlst = new ParamElem[MAXENTITY + 1];
+		mlst[0].code = EOF;
+		pmlst = new ParamElem[MAXENTITY + 1];
+		params = new ParamElem[MAXENTITY + 1];
+		params[0].code = EOF;
+		pparams = new ParamElem[MAXENTITY + 1];
+	}
+
+	if (wrds[wrdidx] == EOF) {
+		wrdidx = 0;
+		scan();
+	} else if (false/*anyOutput*/)	// TODO
+		_vm->paragraph();
+
+	allLength = 0;
+	paramidx = 0;
+	listCopy(pparams, params);
+	params[0].code = EOF;
+	listCopy(pmlst, mlst);
+	mlst[0].code = EOF;
+
+	if (isVerb(wrds[wrdidx])) {
+		vrbwrd = wrds[wrdidx];
+		vrbcode = dict[vrbwrd].code;
+		wrdidx++;
+		match(mlst);
+		action(mlst);		// mlst contains possible multiple params
+	} else {
+	    params[0].code = EOF;
```

Comment:
![image](https://user-images.githubusercontent.com/365188/50493947-c614e480-0a20-11e9-90cf-55730afbd272.png)
 Look at the screenshot, what I see is that there are four spaces instead of tabs.